Display plugin install status, and any installation errors

emlys opened the pull request:
## Description
Fixes #1653 

Moved the spinner into the button, and changed the button text to reflect status of the installation. There are 3-4 different steps:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/e5dac65d-d8cf-4593-8d78-186f1fb1d0cf" />

The "Remove" button is the same, but it just displays "Removing..." because there's only one step.

If there is an error, it now displays the error message and a button to open the log file:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/c440c149-2d18-49d4-8c8e-ec8ce78db1d9" />


## Checklist
- [ ] Updated HISTORY.rst and link to any relevant issue (if these changes are user-facing)
- [ ] Updated the user's guide (if needed)
- [ ] Tested the Workbench UI (if relevant)

dcdenu4 reviewed (approved):
Thanks @emlys , this looks good to me. Some nifty electron communication.

dcdenu4 commented on workbench/src/main/setupAddRemovePlugin.js:
```diff
@@ -71,6 +75,7 @@ export function setupAddPlugin() {
           // Create invest_base environment, if it doesn't already exist
           // The purpose of this environment is just to ensure that git is available
           if (!fs.existsSync(baseEnvPrefix)) {
+            mainWindow.webContents.send('plugin-install-status', i18n.t('Creating base environment...'));
```
Is the behavior the same because we only have one window, mainWindow? But otherwise `e.reply` would be communicating with whatever window sent it?

davemfish replied:
Yeah that sounds right, Doug. Here's the docs on `event.reply`

>reply Function - A function that will send an IPC message to the renderer frame that sent the original message that you are currently handling. You should use this method to "reply" to the sent message in order to guarantee the reply will go to the correct process and frame.
https://www.electronjs.org/docs/latest/api/structures/ipc-main-event

emlys commented:
@davemfish, I tried `event.reply` and got an error that `event.reply is not a function`. At the suggestion of [this stackoverflow thread](https://stackoverflow.com/questions/57038073/ipcmain-event-reply-is-not-a-function) I'm using `event.sender.send`, which works just fine. It's not clear why but I'm guessing that event objects in `handle` are different than event objects in `on`. The other places we use `event.reply`, it's in an `on` handler.

davemfish commented:
> @davemfish, I tried `event.reply` and got an error that `event.reply is not a function`. At the suggestion of [this stackoverflow thread](https://stackoverflow.com/questions/57038073/ipcmain-event-reply-is-not-a-function) I'm using `event.sender.send`, which works just fine. It's not clear why but I'm guessing that event objects in `handle` are different than event objects in `on`. The other places we use `event.reply`, it's in an `on` handler.

Kind of annoying that they are different, and not well-documented. Maybe eventually we will use `handle` for everything, if that's what electron wants us to do.

emlys commented:
@dcdenu4, I realized that it's very similar to display any errors during uninstall, so I added that to the PR (see also https://github.com/natcap/invest/issues/1666. This doesn't cover warnings, just errors). 

Also changed the uninstall command from `mamba remove` to `mamba env remove` - this removes the environment directory as well as the packages in it.

dcdenu4 reviewed (approved):
Thanks @emlys. I'll leave it to @davemfish for the final sign off and merge.

I was curious what your current thinking is for testing plugins? It seems like it'd be good to have tests for the install / uninstall error handling, but I know that getting things working in puppeteer was a blocker early on.

davemfish commented on workbench/tests/renderer/plugin.test.js:
```diff
@@ -73,13 +73,14 @@ describe('Add plugin modal', () => {
-    await waitFor(() => expect(queryByRole('dialog')).toBeNull());
+    // close the modal
+    const overlay = await findByRole('dialog');
+    await act(async () => { await userEvent.click(overlay); });
```
From the look of it, this should work just the same as `await userEvent.click(overlay);`

It looks like there are a few other cases in this file where `userEvent` is sometimes `awaited`, sometimes not, and sometimes wrapped in `act`. I think the cleanest and most reliable pattern is to always do,

`await userEvent.*`

emlys replied:
I was attempting to resolve the `Warning: An update to Transition inside a test was not wrapped in act(...).` messages, but you're right this doesn't seem to be doing anything here. I added `await` to all the `userEvent` calls in the test suite, and that removes the `act()` warnings... even when the warning is coming from a different line of code.

davemfish replied:
Great! I think that's the right approach. My understanding is that even though React issues those `not wrapped in act` warnings, the correct approach for us, since we're using `testing-library` APIs, is basically never to wrap in `act`, but to look for something async that we forgot to `await`.

davemfish commented on workbench/tests/jest.overrides.js:
```diff
@@ -23,6 +23,9 @@ if (global.window) {
   global.window.fetch = fetch;
+
+  global.window.electronAPI = {};
+  global.window.electronAPI.onPluginInstallStatus = () => {};
```
I don't think we're using `window.electronAPI` anymore, right?

davemfish reviewed (changes requested):
Hey @emlys , this looks really good. I left a few small comments that might resolve some warnings you may see when running the tests.

emlys commented on workbench/src/renderer/components/PluginModal/index.jsx:
```diff
@@ -15,52 +15,63 @@ const { ipcRenderer } = window.Workbench.electron;
-  const [loading, setLoading] = useState(false);
+  const [installLoading, setInstallLoading] = useState(false);
+  const [uninstallLoading, setUninstallLoading] = useState(false);
+  const [statusMessage, setStatusMessage] = useState(undefined);
```
Thanks @davemfish. I intended to use strings for all these but clearly I missed some.
